sleap_nn/inference/inference.py

```diff
@@ -165,9 +174,189 @@ def predict(
+                sample_inds=[0] * n,
+            )
+
+            # Access top left point (x,y) of bounding box and subtract this offset from
+            # position of nodes.
+            point = instance_bbox[0, :, 0]
+            centered_centroid = centroid - point
+
+            ex = {}
+            ex["image"] = torch.cat([image] * n)
+            ex["centroid"] = centered_centroid
+            ex["centroid_val"] = centroid_val
+            ex["frame_idx"] = torch.Tensor([fidx] * n)
+            ex["video_idx"] = torch.Tensor([vidx] * n)
+            ex["instance_bbox"] = instance_bbox.squeeze(dim=0).unsqueeze(dim=1)
+            ex["instance_image"] = instance_image.unsqueeze(dim=1)
+            ex["orig_size"] = torch.cat([torch.Tensor(sz)] * n)
+            crops_dict.append(ex)
+
+        return crops_dict
+
+    def forward(self, inputs: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor]:
+        """Predict centroid confidence maps and crop around peaks.
+
+        This layer can be chained with a `FindInstancePeaks` layer to create a top-down
+        inference function from full images.
+
+        Args:
+            inputs: Dictionary with keys:
+                `"image"`: Cropped images.
+                Other keys will be passed down the pipeline.
+
+        Returns:
+            A dictionary of outputs grouped by sample with keys:
+
+            `"centroids"`: The predicted centroids of shape `(batch, max_instances, 2)`.
+            `"centroid_vals": The centroid confidence values of shape `(batch, max_instances)`.
+        """
+        # Network forward pass.
+        cms = self.torch_model(inputs["image"])
+
+        refined_peaks, peak_vals, peak_sample_inds, _ = find_local_peaks(
+            cms.detach(),
+            threshold=self.peak_threshold,
+            refinement=self.refinement,
+            integral_patch_size=self.integral_patch_size,
+        )
+
+        # Adjust for stride and scale.
+        refined_peaks = refined_peaks * self.output_stride  # (n_centroids, 2)
+        batch, channels, height, width = cms.shape
+
+        batch_indxs = Counter(peak_sample_inds.detach().cpu().numpy().tolist())
+
+        self.refined_peaks_with_nans = torch.zeros((batch, self.max_instances, 2))
+        self.peak_vals_with_nans = torch.zeros((batch, self.max_instances))
+
+        parsed = 0
+        for b in range(batch):
+            parsed_idx = batch_indxs[b]
+            current_peaks = refined_peaks[parsed : parsed + parsed_idx]
+            current_peak_vals = peak_vals[parsed : parsed + parsed_idx]
+            if len(current_peaks) > self.max_instances:
+                # take top k
+                current_peak_vals, indices = torch.topk(
+                    current_peak_vals, self.max_instances
+                )
+                current_peaks = current_peaks[indices]
+                num_nans = 0
+            else:
+                num_nans = self.max_instances - len(current_peaks)
```

Comment:
There isn't a way we can do this without forcing max instances?

Reply:
I thought it's better to have max_instance set as we need to process all the tensors as a batch

Reply:
Changed this. max_instances is used only when the user provides or if return_crops = True (when we need to pass it to FindInstancePeaksGroundTruth class)